File: core/src/main/java/com/virtual/box/core/hook/delegate/BaseInstrumentationDelegate.java
package com.virtual.box.core.hook.delegate;


import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Application;
import android.app.Fragment;
import android.app.Instrumentation;
import android.app.UiAutomation;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ActivityInfo;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.os.PersistableBundle;
import android.os.UserHandle;
import android.view.KeyEvent;
import android.view.MotionEvent;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatDelegate;

import com.virtual.box.base.util.log.L;
import com.virtual.box.reflect.MirrorReflection;

public abstract class BaseInstrumentationDelegate extends Instrumentation {

    public static final String TAG = "BaseInstrumentationDelegate";

    protected Instrumentation mBaseInstrumentation;

    /**
     * ActivityThread#handleBindApplication中调用
     */
    @Override
    public void onCreate(Bundle arguments) {
        mBaseInstrumentation.onCreate(arguments);
    }

    @Override
    public void start() {
        mBaseInstrumentation.start();
    }

    @Override
    public void onStart() {
        mBaseInstrumentation.onStart();
    }

    @Override
    public boolean onException(Object obj, Throwable e) {
        return mBaseInstrumentation.onException(obj, e);
    }

    @Override
    public void sendStatus(int resultCode, Bundle results) {
        mBaseInstrumentation.sendStatus(resultCode, results);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    public void addResults(Bundle results) {
        mBaseInstrumentation.addResults(results);
    }

    @Override
    public void setAutomaticPerformanceSnapshots() {
        mBaseInstrumentation.setAutomaticPerformanceSnapshots();
    }

    @Override
    public void startPerformanceSnapshot() {
        mBaseInstrumentation.startPerformanceSnapshot();
    }

    @Override
    public void endPerformanceSnapshot() {
        mBaseInstrumentation.endPerformanceSnapshot();
    }

    @Override
    public void onDestroy() {
        mBaseInstrumentation.onDestroy();
    }

    @Override
    public Context getContext() {
        Context context =  mBaseInstrumentation.getContext();
        L.hd("[%s] >> getContext >> result = %s", TAG, context);
        return context;
    }

    @Override
    public ComponentName getComponentName() {
        ComponentName componentName = mBaseInstrumentation.getComponentName();
        L.hd("[%s] >> getComponentName >> result = %s", TAG, componentName);
        return componentName;
    }

    @Override
    public Context getTargetContext() {
        Context context = mBaseInstrumentation.getTargetContext();
        L.hd("[%s] >> getTargetContext >> result = %s", TAG, context);
        return context;
    }

    @Override
    public boolean isProfiling() {
        return mBaseInstrumentation.isProfiling();
    }

    @Override
    public void startProfiling() {
        mBaseInstrumentation.startProfiling();
    }

    @Override
    public void stopProfiling() {
        mBaseInstrumentation.stopProfiling();
    }

    @Override
    public void setInTouchMode(boolean inTouch) {
        mBaseInstrumentation.setInTouchMode(inTouch);
    }

    @Override
    public void waitForIdle(Runnable recipient) {
        mBaseInstrumentation.waitForIdle(recipient);
    }

    @Override
    public void waitForIdleSync() {
        mBaseInstrumentation.waitForIdleSync();
    }

    @Override
    public void runOnMainSync(Runnable runner) {
        mBaseInstrumentation.runOnMainSync(runner);
    }

    @Override
    public Activity startActivitySync(Intent intent) {
        return mBaseInstrumentation.startActivitySync(intent);
    }

    @Override
    public void addMonitor(ActivityMonitor monitor) {
        mBaseInstrumentation.addMonitor(monitor);
    }

    @Override
    public ActivityMonitor addMonitor(IntentFilter filter, ActivityResult result, boolean block) {
        return mBaseInstrumentation.addMonitor(filter, result, block);
    }

    @Override
    public ActivityMonitor addMonitor(String cls, ActivityResult result, boolean block) {
        return mBaseInstrumentation.addMonitor(cls, result, block);
    }

    @Override
    public boolean checkMonitorHit(ActivityMonitor monitor, int minHits) {
        return mBaseInstrumentation.checkMonitorHit(monitor, minHits);
    }

    @Override
    public Activity waitForMonitor(ActivityMonitor monitor) {
        return mBaseInstrumentation.waitForMonitor(monitor);
    }

    @Override
    public Activity waitForMonitorWithTimeout(ActivityMonitor monitor, long timeOut) {
        return mBaseInstrumentation.waitForMonitorWithTimeout(monitor, timeOut);
    }

    @Override
    public void removeMonitor(ActivityMonitor monitor) {
        mBaseInstrumentation.removeMonitor(monitor);
    }

    @Override
    public boolean invokeMenuActionSync(Activity targetActivity, int id, int flag) {
        return mBaseInstrumentation.invokeMenuActionSync(targetActivity, id, flag);
    }

    @Override
    public boolean invokeContextMenuAction(Activity targetActivity, int id, int flag) {
        return mBaseInstrumentation.invokeContextMenuAction(targetActivity, id, flag);
    }

    @Override
    public void sendStringSync(String text) {
        mBaseInstrumentation.sendStringSync(text);
    }

    @Override
    public void sendKeySync(KeyEvent event) {
        mBaseInstrumentation.sendKeySync(event);
    }

    @Override
    public void sendKeyDownUpSync(int key) {
        mBaseInstrumentation.sendKeyDownUpSync(key);
    }

    @Override
    public void sendCharacterSync(int keyCode) {
        mBaseInstrumentation.sendCharacterSync(keyCode);
    }

    @Override
    public void sendPointerSync(MotionEvent event) {
        mBaseInstrumentation.sendPointerSync(event);
    }

    @Override
    public void sendTrackballEventSync(MotionEvent event) {
        mBaseInstrumentation.sendTrackballEventSync(event);
    }

    @Override
    public Application newApplication(ClassLoader cl, String className, Context context) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
        Application application = mBaseInstrumentation.newApplication(cl, className, context);
        L.hdParamTag(TAG, "cl = %s, className = %s, context = %s, result = %s",cl, className, context, application);
        //Debug.waitForDebugger();
        return application;
    }

    @Override
    public void callApplicationOnCreate(Application app) {
        L.hd("[%s] >> callApplicationOnCreate >> app = %s", TAG, app);
        mBaseInstrumentation.callApplicationOnCreate(app);
    }

    @Override
    public Activity newActivity(Class<?> clazz, Context context, IBinder token, Application application, Intent intent,
                                ActivityInfo info, CharSequence title, Activity parent, String id, Object lastNonConfigurationInstance) throws IllegalAccessException, InstantiationException {
        Activity activity =  mBaseInstrumentation.newActivity(clazz, context, token, application, intent, info, title, parent, id, lastNonConfigurationInstance);
        L.hdParamTag(TAG,"clazz = %s, context = %s, token = %s, application = %s, intent = %s, info = %s, title = %s, parent = %s, id = %s," +
                        " lastNonConfigurationInstance = %s",
                clazz, context, token, application, intent, info, title, parent, id, lastNonConfigurationInstance);
        return activity;
    }

    @Override
    public Activity newActivity(ClassLoader cl, String className, Intent intent) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
        Activity activity =  mBaseInstrumentation.newActivity(cl, className, intent);
        L.hdParamTag(TAG,"classLoader = %s, className = %s, intent = %s, result = %s", cl, className, intent, activity);
        return activity;
    }

    @Override
    public void callActivityOnCreate(Activity activity, Bundle icicle) {
//        Debug.waitForDebugger();
        L.hdParamTag(TAG,"activity = %s, bundle = %s", activity, icicle);
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
        mBaseInstrumentation.callActivityOnCreate(activity, icicle);
    }

    @Override
    public void callActivityOnCreate(Activity activity, Bundle icicle, PersistableBundle persistentState) {
        mBaseInstrumentation.callActivityOnCreate(activity, icicle, persistentState);
    }

    @Override
    public void callActivityOnDestroy(Activity activity) {
        L.hdParamTag(TAG,"activity = %s", activity);
        mBaseInstrumentation.callActivityOnDestroy(activity);
        // 销毁后如果如果没有Activity存在了，就把当前进程干掉
    }

    @Override
    public void callActivityOnRestoreInstanceState(Activity activity, Bundle savedInstanceState) {
        mBaseInstrumentation.callActivityOnRestoreInstanceState(activity, savedInstanceState);
    }

    @Override
    public void callActivityOnRestoreInstanceState(Activity activity, Bundle savedInstanceState, PersistableBundle persistentState) {
        mBaseInstrumentation.callActivityOnRestoreInstanceState(activity, savedInstanceState, persistentState);
    }

    @Override
    public void callActivityOnPostCreate(Activity activity, Bundle icicle) {
        mBaseInstrumentation.callActivityOnPostCreate(activity, icicle);
    }

    @Override
    public void callActivityOnPostCreate(Activity activity, Bundle icicle, PersistableBundle persistentState) {
        mBaseInstrumentation.callActivityOnPostCreate(activity, icicle, persistentState);
    }

    @Override
    public void callActivityOnNewIntent(Activity activity, Intent intent) {
        mBaseInstrumentation.callActivityOnNewIntent(activity, intent);
    }

    @Override
    public void callActivityOnStart(Activity activity) {
        mBaseInstrumentation.callActivityOnStart(activity);
    }

    @Override
    public void callActivityOnRestart(Activity activity) {
        mBaseInstrumentation.callActivityOnRestart(activity);
    }

    @Override
    public void callActivityOnResume(Activity activity) {
        mBaseInstrumentation. callActivityOnResume(activity);
    }

    @Override
    public void callActivityOnStop(Activity activity) {
        mBaseInstrumentation.callActivityOnStop(activity);
    }

    @Override
    public void callActivityOnSaveInstanceState(Activity activity, Bundle outState) {
        mBaseInstrumentation.callActivityOnSaveInstanceState(activity, outState);
    }

    @Override
    public void callActivityOnSaveInstanceState(Activity activity, Bundle outState, PersistableBundle outPersistentState) {
        mBaseInstrumentation.callActivityOnSaveInstanceState(activity, outState, outPersistentState);
    }

    @Override
    public void callActivityOnPause(Activity activity) {
        mBaseInstrumentation.callActivityOnPause(activity);
    }

    @Override
    public void callActivityOnUserLeaving(Activity activity) {
        mBaseInstrumentation.callActivityOnUserLeaving(activity);
    }

    @Override
    public void startAllocCounting() {
        mBaseInstrumentation.startAllocCounting();
    }

    @Override
    public void stopAllocCounting() {
        mBaseInstrumentation.stopAllocCounting();
    }

    @Override
    public Bundle getAllocCounts() {
        return mBaseInstrumentation.getAllocCounts();
    }

    @Override
    public Bundle getBinderCounts() {
        return mBaseInstrumentation.getBinderCounts();
    }

    @Override
    public UiAutomation getUiAutomation() {
        return mBaseInstrumentation.getUiAutomation();
    }

    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, Activity activity, Intent intent, int requestCode, Bundle options) throws Throwable {
        return invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                Activity.class,
                Intent.class,
                Integer.TYPE,
                Bundle.class).call(mBaseInstrumentation, context, contextThread, token, activity, intent, requestCode, options);
    }

    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, String str, Intent intent, int requestCode, Bundle options) throws Throwable {
        return invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                String.class,
                Intent.class,
                Integer.TYPE,
                Bundle.class).call(mBaseInstrumentation, context, contextThread, token, str, intent, requestCode, options);
    }

    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, Fragment fragment, Intent intent, int requestCode) throws Throwable {
        return invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                Fragment.class,
                Intent.class,
                Integer.TYPE).call(mBaseInstrumentation, context, contextThread, token, fragment, intent, requestCode);
    }

    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, Activity activity, Intent intent, int requestCode) throws Throwable {
        return invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                Activity.class,
                Intent.class,
                Integer.TYPE).call(mBaseInstrumentation, context, contextThread, token, activity, intent, requestCode);
    }

    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, Fragment fragment, Intent intent, int requestCode, Bundle bundle) throws Throwable {
        return invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                Fragment.class,
                Intent.class,
                Integer.TYPE,
                Bundle.class).call(mBaseInstrumentation, context, contextThread, token, fragment, intent, requestCode, bundle);
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    public ActivityResult execStartActivity(Context context, IBinder contextThread, IBinder token, Activity activity, Intent intent, int requestCode, Bundle bundle, UserHandle userHandle) throws Throwable {
        return (ActivityResult) invokeExecStartActivity(mBaseInstrumentation,
                Context.class,
                IBinder.class,
                IBinder.class,
                Activity.class,
                Intent.class,
                Integer.TYPE,
                Bundle.class,
                UserHandle.class).call(mBaseInstrumentation, new Object[]{context, contextThread, token, activity, intent, requestCode, bundle, userHandle});
    }

    private static MirrorReflection.MethodWrapper<ActivityResult> invokeExecStartActivity(Object obj, Class<?>... args) throws NoSuchMethodException {
        Class<?> cls = obj.getClass();
        while (cls != null) {
            try {
                return MirrorReflection.on(obj.getClass())
                        .method("execStartActivity", args);
            } catch (Exception e) {
                cls = cls.getSuperclass();
            }
        }
        throw new NoSuchMethodException();
    }
}

File: core/src/main/java/com/virtual/box/core/hook/delegate/BaseInstrumentationDelegate.kt
package com.virtual.box.core.hook.delegate

import android.annotation.TargetApi
import android.app.*
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ActivityInfo
import android.os.*
import android.view.KeyEvent
import android.view.MotionEvent
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatDelegate
import com.virtual.box.base.util.compat.BuildCompat
import com.virtual.box.base.util.log.L
import com.virtual.box.base.util.log.Logger.Companion.getLogger
import com.virtual.box.core.compat.ComponentFixCompat
import com.virtual.box.reflect.MirrorReflection
import com.virtual.box.reflect.MirrorReflection.MethodWrapper
import com.virtual.box.reflect.android.app.HActivity
import com.virtual.box.reflect.android.app.HApplication
import com.virtual.box.reflect.android.content.pm.HActivityInfo

abstract class BaseInstrumentationDelegate : Instrumentation() {
    protected var mBaseInstrumentation: Instrumentation? = null

    /**
     * ActivityThread#handleBindApplication中调用
     */
    override fun onCreate(arguments: Bundle?) {
        mBaseInstrumentation?.onCreate(arguments)
    }

    override fun start() {
        mBaseInstrumentation?.start()
    }

    override fun onStart() {
        mBaseInstrumentation?.onStart()
    }

    override fun onException(obj: Any?, e: Throwable?): Boolean {
        logger.e("onException#obj = %s, e = %s", obj, e)
        return mBaseInstrumentation!!.onException(obj, e)
    }

    override fun sendStatus(resultCode: Int, results: Bundle) {
        mBaseInstrumentation?.sendStatus(resultCode, results)
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    override fun addResults(results: Bundle) {
        mBaseInstrumentation?.addResults(results)
    }

    override fun setAutomaticPerformanceSnapshots() {
        mBaseInstrumentation?.setAutomaticPerformanceSnapshots()
    }

    override fun startPerformanceSnapshot() {
        mBaseInstrumentation?.startPerformanceSnapshot()
    }

    override fun endPerformanceSnapshot() {
        mBaseInstrumentation?.endPerformanceSnapshot()
    }

    override fun onDestroy() {
        mBaseInstrumentation?.onDestroy()
    }

    override fun getContext(): Context {
        val context = mBaseInstrumentation!!.context
        logger.i("getContext#result = %s", context)
        return context
    }

    override fun getComponentName(): ComponentName {
        val componentName = mBaseInstrumentation!!.componentName
        logger.i("getComponentName#result = %s", componentName)
        return componentName
    }

    override fun getTargetContext(): Context {
        val context = mBaseInstrumentation!!.targetContext
        logger.i("getTargetContext#result = %s", context)
        return context
    }

    override fun isProfiling(): Boolean {
        val result = mBaseInstrumentation!!.isProfiling
        logger.i("isProfiling#result = %s", result)
        return result
    }

    override fun startProfiling() {
        mBaseInstrumentation?.startProfiling()
    }

    override fun stopProfiling() {
        mBaseInstrumentation?.stopProfiling()
    }

    override fun setInTouchMode(inTouch: Boolean) {
        mBaseInstrumentation?.setInTouchMode(inTouch)
    }

    override fun waitForIdle(recipient: Runnable) {
        logger.i("waitForIdle")
        mBaseInstrumentation?.waitForIdle(recipient)
    }

    override fun waitForIdleSync() {
        logger.i("waitForIdleSync")
        mBaseInstrumentation?.waitForIdleSync()
    }

    override fun runOnMainSync(runner: Runnable) {
        logger.i("runOnMainSync")
        mBaseInstrumentation?.runOnMainSync(runner)
    }

    override fun startActivitySync(intent: Intent): Activity {
        logger.i("startActivitySync#intent = %s", intent)
        return mBaseInstrumentation!!.startActivitySync(intent)
    }

    override fun addMonitor(monitor: ActivityMonitor) {
        mBaseInstrumentation?.addMonitor(monitor)
    }

    override fun addMonitor(filter: IntentFilter, result: ActivityResult, block: Boolean): ActivityMonitor {
        return mBaseInstrumentation!!.addMonitor(filter, result, block)
    }

    override fun addMonitor(cls: String, result: ActivityResult, block: Boolean): ActivityMonitor {
        return mBaseInstrumentation!!.addMonitor(cls, result, block)
    }

    override fun checkMonitorHit(monitor: ActivityMonitor, minHits: Int): Boolean {
        return mBaseInstrumentation!!.checkMonitorHit(monitor, minHits)
    }

    override fun waitForMonitor(monitor: ActivityMonitor): Activity {
        return mBaseInstrumentation!!.waitForMonitor(monitor)
    }

    override fun waitForMonitorWithTimeout(monitor: ActivityMonitor, timeOut: Long): Activity {
        logger.i("waitForMonitorWithTimeout#monitor = %s, timeOut = %s", monitor, timeOut)
        return mBaseInstrumentation!!.waitForMonitorWithTimeout(monitor, timeOut)
    }

    override fun removeMonitor(monitor: ActivityMonitor) {
        mBaseInstrumentation?.removeMonitor(monitor)
    }

    override fun invokeMenuActionSync(targetActivity: Activity, id: Int, flag: Int): Boolean {
        return mBaseInstrumentation!!.invokeMenuActionSync(targetActivity, id, flag)
    }

    override fun invokeContextMenuAction(targetActivity: Activity, id: Int, flag: Int): Boolean {
        return mBaseInstrumentation!!.invokeContextMenuAction(targetActivity, id, flag)
    }

    override fun sendStringSync(text: String) {
        mBaseInstrumentation?.sendStringSync(text)
    }

    override fun sendKeySync(event: KeyEvent) {
        mBaseInstrumentation?.sendKeySync(event)
    }

    override fun sendKeyDownUpSync(key: Int) {
        mBaseInstrumentation?.sendKeyDownUpSync(key)
    }

    override fun sendCharacterSync(keyCode: Int) {
        mBaseInstrumentation?.sendCharacterSync(keyCode)
    }

    override fun sendPointerSync(event: MotionEvent) {
        mBaseInstrumentation?.sendPointerSync(event)
    }

    override fun sendTrackballEventSync(event: MotionEvent) {
        mBaseInstrumentation?.sendTrackballEventSync(event)
    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InstantiationException::class)
    override fun newApplication(cl: ClassLoader, className: String, context: Context): Application {
        logger.i("newApplication#cl = %s, className = %s, context = %s", cl, className, context)
        val application = mBaseInstrumentation!!.newApplication(cl, className, context)
        logger.i("newApplication#result = %s", application)
        //Debug.waitForDebugger();
        return application
    }

    override fun callApplicationOnCreate(app: Application?) {
        logger.i("callApplicationOnCreate#app = %s", app)
        mBaseInstrumentation?.callApplicationOnCreate(app)
    }

    @Throws(IllegalAccessException::class, InstantiationException::class)
    override fun newActivity(
        clazz: Class<*>?, context: Context, token: IBinder, application: Application, intent: Intent,
        info: ActivityInfo, title: CharSequence, parent: Activity, id: String, lastNonConfigurationInstance: Any
    ): Activity {
        logger.i(
            "newActivity#class = %s, context = %s, token = %s, app = %s, intent = %s, info = %s, title = %s, parent = %s, id = %s," +
                    " lastNonConfigurationInstance = %s",
            clazz, context, token, application, intent, info, title, parent, id, lastNonConfigurationInstance
        )
        val activity = mBaseInstrumentation!!.newActivity(clazz, context, token, application, intent,
            info, title, parent, id, lastNonConfigurationInstance
        )
        logger.i("newActivity#result = %s", activity)
        return activity
    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InstantiationException::class)
    override fun newActivity(cl: ClassLoader, className: String, intent: Intent): Activity {
        logger.i("newActivity#classLoader = %s, className = %s, intent = %s", cl, className, intent)
        val activity = mBaseInstrumentation!!.newActivity(cl, className, intent)
        logger.i("newActivityInfo# result = %s", activity)
        return activity
    }

    override fun callActivityOnCreate(activity: Activity, icicle: Bundle?) {
        logger.i("callActivityOnCreate#activity = %s, bundle = %s", activity, icicle)
        if (BuildCompat.isAtLeastS){
            ComponentFixCompat.fixActivityWithOnCreate(activity)
        }
        val info = HActivity.mActivityInfo.get(activity)
        if (info.theme != 0) {
            activity.theme.applyStyle(info.theme, true)
        }else{
            val appInfo = HActivityInfo.applicationInfo.get(info)
            if (appInfo.theme != 0){
                activity.theme.applyStyle(appInfo.theme, true)
            }
        }
        ComponentFixCompat.fixActivityOrientation(activity)
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        mBaseInstrumentation?.callActivityOnCreate(activity, icicle)
    }

    override fun callActivityOnCreate(activity: Activity, icicle: Bundle?, persistentState: PersistableBundle?) {
        logger.i("callActivityOnCreate#activity = %s, bundle = %s, persistentState = %s", activity, icicle, persistentState)
        mBaseInstrumentation?.callActivityOnCreate(activity, icicle, persistentState)
    }

    override fun callActivityOnDestroy(activity: Activity) {
        logger.i("callActivityOnDestroy#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnDestroy(activity)
        // 销毁后如果如果没有Activity存在了，就把当前进程干掉
    }

    override fun callActivityOnRestoreInstanceState(activity: Activity, savedInstanceState: Bundle) {
        logger.i("callActivityOnRestoreInstanceState#activity = %s, savedInstanceState = %s", activity, savedInstanceState)
        mBaseInstrumentation?.callActivityOnRestoreInstanceState(activity, savedInstanceState)
    }

    override fun callActivityOnRestoreInstanceState(activity: Activity, savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        logger.i(
            "callActivityOnRestoreInstanceState#activity = %s, savedInstanceState = %s, persistentState = %s",
            activity, savedInstanceState, persistentState
        )
        mBaseInstrumentation?.callActivityOnRestoreInstanceState(activity, savedInstanceState, persistentState)
    }

    override fun callActivityOnPostCreate(activity: Activity, icicle: Bundle?) {
        logger.i("callActivityOnPostCreate#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnPostCreate(activity, icicle)
    }

    override fun callActivityOnPostCreate(activity: Activity, icicle: Bundle?, persistentState: PersistableBundle?) {
        logger.i("callActivityOnPostCreate#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnPostCreate(activity, icicle, persistentState)
    }

    override fun callActivityOnNewIntent(activity: Activity?, intent: Intent?) {
        logger.i("callActivityOnNewIntent#activity = %s, intent = %s", activity, intent)
        mBaseInstrumentation?.callActivityOnNewIntent(activity, intent)
    }

    override fun callActivityOnStart(activity: Activity) {
        logger.i("callActivityOnStart#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnStart(activity)
    }

    override fun callActivityOnRestart(activity: Activity) {
        logger.i("callActivityOnReStart#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnRestart(activity)
    }

    override fun callActivityOnResume(activity: Activity) {
        logger.i("callActivityOnResume#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnResume(activity)
    }

    override fun callActivityOnStop(activity: Activity) {
        logger.i("callActivityOnStop#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnStop(activity)
    }

    override fun callActivityOnSaveInstanceState(activity: Activity, outState: Bundle) {
        logger.i("callActivityOnSaveInstanceState#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnSaveInstanceState(activity, outState)
    }

    override fun callActivityOnSaveInstanceState(activity: Activity, outState: Bundle, outPersistentState: PersistableBundle) {
        logger.i("callActivityOnSaveInstanceState#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnSaveInstanceState(activity, outState, outPersistentState)
    }

    override fun callActivityOnPause(activity: Activity) {
        logger.i("callActivityOnPause#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnPause(activity)
    }

    override fun callActivityOnUserLeaving(activity: Activity) {
        logger.i("callActivityOnUserLeaving#activity = %s", activity)
        mBaseInstrumentation?.callActivityOnUserLeaving(activity)
    }

    override fun startAllocCounting() {
        mBaseInstrumentation?.startAllocCounting()
    }

    override fun stopAllocCounting() {
        mBaseInstrumentation?.stopAllocCounting()
    }

    override fun getAllocCounts(): Bundle {
        return mBaseInstrumentation!!.allocCounts
    }

    override fun getBinderCounts(): Bundle {
        return mBaseInstrumentation!!.binderCounts
    }

    override fun getUiAutomation(): UiAutomation {
        return mBaseInstrumentation!!.uiAutomation
    }

    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context, contextThread: IBinder,
        token: IBinder?, activity: Activity?, intent: Intent,
        requestCode: Int, options: Bundle?
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            Activity::class.java,
            Intent::class.java,
            Integer.TYPE,
            Bundle::class.java
        ).call(mBaseInstrumentation, context, contextThread, token, activity, intent, requestCode, options)
    }

    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context,
        contextThread: IBinder,
        token: IBinder?,
        str: String?,
        intent: Intent,
        requestCode: Int,
        options: Bundle?
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            String::class.java,
            Intent::class.java,
            Integer.TYPE,
            Bundle::class.java
        ).call(mBaseInstrumentation, context, contextThread, token, str, intent, requestCode, options)
    }

    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context,
        contextThread: IBinder,
        token: IBinder?,
        fragment: Fragment?,
        intent: Intent,
        requestCode: Int
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            Fragment::class.java,
            Intent::class.java,
            Integer.TYPE
        ).call(mBaseInstrumentation, context, contextThread, token, fragment, intent, requestCode)
    }

    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context,
        contextThread: IBinder,
        token: IBinder?,
        activity: Activity?,
        intent: Intent,
        requestCode: Int
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            Activity::class.java,
            Intent::class.java,
            Integer.TYPE
        ).call(mBaseInstrumentation, context, contextThread, token, activity, intent, requestCode)
    }

    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context,
        contextThread: IBinder,
        token: IBinder?,
        fragment: Fragment?,
        intent: Intent,
        requestCode: Int,
        bundle: Bundle?
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            Fragment::class.java,
            Intent::class.java,
            Integer.TYPE,
            Bundle::class.java
        ).call(mBaseInstrumentation, context, contextThread, token, fragment, intent, requestCode, bundle)
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    @Throws(Throwable::class)
    open fun execStartActivity(
        context: Context,
        contextThread: IBinder,
        token: IBinder?,
        activity: Activity?,
        intent: Intent,
        requestCode: Int,
        bundle: Bundle?,
        userHandle: UserHandle?
    ): ActivityResult? {
        return invokeExecStartActivity(
            mBaseInstrumentation!!,
            Context::class.java,
            IBinder::class.java,
            IBinder::class.java,
            Activity::class.java,
            Intent::class.java,
            Integer.TYPE,
            Bundle::class.java,
            UserHandle::class.java
        ).call(
            mBaseInstrumentation,
            *arrayOf(context, contextThread, token, activity, intent, requestCode, bundle, userHandle)
        ) as ActivityResult
    }

    companion object {
        protected val logger = getLogger(L.HOOK_TAG, "InstrumentationDelegate")
        const val TAG = "BaseInstrumentationDelegate"
        @Throws(NoSuchMethodException::class)
        private fun invokeExecStartActivity(obj: Any, vararg args: Class<*>): MethodWrapper<ActivityResult> {
            var cls: Class<*>? = obj.javaClass
            while (cls != null) {
                cls = try {
                    return MirrorReflection.on(obj.javaClass)
                        .method("execStartActivity", *args)
                } catch (e: Exception) {
                    cls.superclass
                }
            }
            throw NoSuchMethodException()
        }
    }
}